core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java

```diff
@@ -119,11 +130,74 @@ public static Decision tryShardOnNode(ShardRouting shard, RoutingNode node, Rout
+        ClusterAllocationExplanation.FinalDecision finalDecision;
+        ClusterAllocationExplanation.StoreCopy storeCopy;
+        String finalExplanation;
+
+        if (node.getId().equals(assignedNodeId)) {
+            finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
+            finalExplanation = "the shard is already assigned to this node";
+        } else if (nodeDecision.type() == Decision.Type.NO) {
+            finalDecision = ClusterAllocationExplanation.FinalDecision.NO;
+            finalExplanation = "the shard cannot be assigned because one or more allocation decider returns a 'NO' decision";
+        } else {
+            finalDecision = ClusterAllocationExplanation.FinalDecision.YES;
+            finalExplanation = "the shard can be assigned";
+        }
+
+        if (storeStatus != null) {
+            final Throwable storeErr = storeStatus.getStoreException();
+            // The store error only influences the decision if the shard is primary and has not been allocated before
+            if (storeErr != null && shard.primary() && shard.allocatedPostIndexCreate(indexMetaData) == false) {
+                finalDecision = ClusterAllocationExplanation.FinalDecision.NO;
+                if (ExceptionsHelper.unwrapCause(storeErr) instanceof CorruptIndexException) {
+                    storeCopy = ClusterAllocationExplanation.StoreCopy.CORRUPT;
+                    finalExplanation = "the copy of data in the shard store is corrupt";
+                } else {
+                    storeCopy = ClusterAllocationExplanation.StoreCopy.IO_ERROR;
+                    finalExplanation = "there was an IO error reading from data in the shard store";
+                }
+            } else if (activeAllocationIds.isEmpty()) {
+                // The ids are only empty if dealing with a legacy index
+                // TODO: fetch the shard state versions and display here?
+                storeCopy = ClusterAllocationExplanation.StoreCopy.UNKNOWN;
+            } else if (activeAllocationIds.contains(storeStatus.getAllocationId())) {
+                storeCopy = ClusterAllocationExplanation.StoreCopy.AVAILABLE;
+                if (finalDecision == ClusterAllocationExplanation.FinalDecision.YES) {
+                    finalExplanation = "the shard can be assigned and the node contains a valid copy of the shard data";
+                }
+            } else {
+                // Otherwise, this is a stale copy of the data (allocation ids don't match)
+                storeCopy = ClusterAllocationExplanation.StoreCopy.STALE;
+                finalExplanation = "the copy of the shard is stale, allocation ids do not match";
+                finalDecision = ClusterAllocationExplanation.FinalDecision.NO;
+            }
+        } else {
+            // No copies of the data, so deciders are what influence the decision and explanation
+            storeCopy = ClusterAllocationExplanation.StoreCopy.NONE;
+        }
```

Comment:
I think this method is not quite right yet. A few observations:
- if (unassigned) primary, then finalExplanation / finalDecision should be influenced by staleness of copy. In case of a replica, however, staleness does not influence allocation decision.
- if replica, then we can still show that store copy is corrupt / has an io / error without influencing final decision / explanation
- for the primary / replica shard allocator, corrupt / io_error is treated as no data.

I think we can first calculate store copy (which just represents the status on disk) and then influence finalExpl / finalDecision based on that.

Here is my go at it:

```
    public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
                                                           DiscoveryNode node,
                                                           Decision nodeDecision,
                                                           Float nodeWeight,
                                                           IndicesShardStoresResponse.StoreStatus storeStatus,
                                                           String assignedNodeId,
                                                           Set<String> activeAllocationIds) {

        final StoreCopy storeCopy;

        if (storeStatus == null) {
            // No copies of the data
            storeCopy = StoreCopy.NONE;
        } else {
            final Throwable storeErr = storeStatus.getStoreException();
            if (storeErr != null) {
                if (ExceptionsHelper.unwrapCause(storeErr) instanceof CorruptIndexException) {
                    storeCopy = StoreCopy.CORRUPT;
                } else {
                    storeCopy = StoreCopy.IO_ERROR;
                }
            } else if (activeAllocationIds.isEmpty()) {
                // The ids are only empty if dealing with a legacy index
                // TODO: fetch the shard state versions and display here?
                storeCopy = StoreCopy.UNKNOWN;
            } else if (activeAllocationIds.contains(storeStatus.getAllocationId())) {
                storeCopy = StoreCopy.AVAILABLE;
            } else {
                // Otherwise, this is a stale copy of the data (allocation ids don't match)
                storeCopy = StoreCopy.STALE;
            }
        }

        final FinalDecision finalDecision;
        final String finalExplanation;

        if (node.getId().equals(assignedNodeId)) {
            finalDecision = FinalDecision.ALREADY_ASSIGNED;
            finalExplanation = "the shard is already assigned to this node";
        } else if (shard.primary() && shard.unassigned() && storeCopy == StoreCopy.STALE) {
            finalExplanation = "the copy of the shard is stale, allocation ids do not match";
            finalDecision = FinalDecision.NO;
        } else {
            if (nodeDecision.type() == Decision.Type.NO) {
                finalDecision = FinalDecision.NO;
                finalExplanation = "the shard cannot be assigned because one or more allocation decider returns a 'NO' decision";
            } else {
                finalDecision = FinalDecision.YES;
                if (storeCopy == StoreCopy.AVAILABLE) {
                    finalExplanation = "the shard can be assigned and the node contains a valid copy of the shard data";
                } else {
                    finalExplanation = "the shard can be assigned";
                }
            }
        }

        return new NodeExplanation(node, nodeDecision, nodeWeight, storeStatus, finalDecision, finalExplanation, storeCopy);
    }
```